render/src/lib/views/sidebar.vertical/port.available/components.ts

```diff
@@ -0,0 +1,344 @@
+// tslint:disable:no-inferrable-types
+
+import { Component, OnDestroy, Input, AfterViewInit, ViewChild, ElementRef, ChangeDetectorRef, OnInit } from '@angular/core';
+import { IPortInfo } from '../../../common/interface.portinfo';
+import Chart from 'chart.js';
+import { Subscription, Subject } from 'rxjs';
+import Service, { IPort } from '../../../services/service';
+import { EHostEvents } from '../../../common/host.events';
+import { SidebarVerticalPortDialogComponent } from '../port.options/component';
+import { IOptions, CDefaultOptions } from '../../../common/interface.options';
+import { ENotificationType } from 'chipmunk.client.toolkit';
+import { SidebarVerticalPortWarningComponent } from '../port.warning/component';
+
+const LIMIT = 500;
+
+@Component({
+    selector: 'lib-dia-port-available-com',
+    templateUrl: './template.html',
+    styleUrls: ['./styles.less']
+})
+
+export class DialogAvailablePortComponent implements OnDestroy, AfterViewInit, OnInit {
+
+    @Input() port: IPortInfo;
+    @Input() observables: {tick: Subject<boolean>, resize: Subject<{ sidebar_width: number, sidebar_height: number }>};
+    @ViewChild('canvas') canvas: ElementRef<HTMLCanvasElement>;
+
+    private _subscriptions: { [key: string]: Subscription } = {};
+    private _ctx: any;
+    private _read: number = 0;
+    private _chart: Chart;
+    private _chart_labels = new Array(LIMIT).fill('');
+    private _destroyed: boolean = false;
+    private _options: IOptions;
+    private _defaultOptions: IOptions = Object.assign({}, CDefaultOptions);
+    private _sidebar_width: number;
+    private _session: string;
+
+    public _ng_isAvailable: boolean;
+    public _ng_isConnected: boolean = false;
+
+    constructor(private _cdRef: ChangeDetectorRef) {
+    }
+
+    ngAfterViewInit() {
+        if (Service.getPortAvailable(this.port.path)) {
+            this._createChart();
+            this._loadSession();
+        }
+        this._subscribe();
+    }
+
+    ngOnInit() {
+        this._session = Service.getSessionID();
+        this._defaultOptions.path = this.port.path;
+        this._ng_isAvailable = Service.getPortAvailable(this.port.path);
+    }
+
+    ngOnDestroy() {
+        if (this._chart) {
+            this._chart.destroy();
+            this._chart = undefined;
+        }
+        this._unsubscribe();
+        this._destroyed = true;
+    }
+
+    private _forceUpdate() {
+        if (this._destroyed) {
+            return;
+        }
+        this._cdRef.detectChanges();
+    }
+
+    private _subscribe() {
+        if (Service.getPortAvailable(this.port.path)) {
+            this._subscribeToTick('tick');
+            this._subscribeToResize('resize');
+        }
+        this._subscribeToEvent('event');
+    }
+
+    private _subscribeToTick(key: string) {
```

Comment:
if you have some list of possible `string` values, probably you want to define it like it:
```
private _subscribeToTick(key: 'event_a' | 'event_b' | 'event_c') { ... }
```

Reply:
The key is only there to create an entry for the hashmap, it does not represent the events. The events are in the next line with size: any